packages/dds/tree/src/core/rebase/rebaser.ts

```diff
@@ -3,223 +3,140 @@
-		{ before: RevisionTag; change: ChangesetFromChangeRebaser<TChangeRebaser> }
-	> = new Map();
-
-	public constructor(public readonly rebaser: TChangeRebaser) {
-		// TODO
-	}
-
-	/**
-	 * Rebase `changes` from being applied to the `from` state to able to be applied to the `to` state.
-	 * @returns a RevisionTag for the state after applying changes to `to`, and the rebased changes themselves.
-	 */
-	public rebase(
-		changes: ChangesetFromChangeRebaser<TChangeRebaser>,
-		from: RevisionTag,
-		to: RevisionTag,
-	): [RevisionTag, ChangesetFromChangeRebaser<TChangeRebaser>] {
-		const over = this.getResolutionPath(from, to);
-		const finalChangeset: ChangesetFromChangeRebaser<TChangeRebaser> = this.rebaser.rebase(
-			changes,
-			over,
-		);
-		const newRevision = this.makeRevision();
-		this.revisionTree.set(newRevision, {
-			before: to,
-			change: finalChangeset,
-		});
-		return [newRevision, finalChangeset];
-	}
+	public constructor(public readonly changeRebaser: ChangeRebaser<TChange>) {}
 
 	/**
-	 * Modifies `anchors` to be valid at the destination.
+	 * Rebases a branch (the source) onto another branch (the target).
+	 *
+	 * Given a commit as the "head", a branch is defined as the ancestry path of that head commit.
+	 * The source and target branch must share an ancestor. Respectively, they must not contain any commits
+	 * in their path with duplicate tags, or this function will have undefined behavior. If there are any commits
+	 * in `source` that have the same tag as any commits in `target`, then this function will treat each pair as
+	 * semantically equivalent (i.e. one is a pre-existing rebase of the other) and will attempt to minimize the amount
+	 * of rebasing performed for those changes across the branches.
+	 * @param source - the head of a branch
+	 * @param target - the head of branch to rebase `source` onto
+	 * @param base - the commit in the target branch up to which to perform the rebase. Defaults to `target`. Source will be
+	 * rebased up to at least this commit, but it may rebase farther if it is trivial to do so.
+	 * @returns the head of a rebased source branch and the cumulative change to the source branch
 	 */
-	public rebaseAnchors(anchors: AnchorSet, from: RevisionTag, to: RevisionTag): void {
-		const over = this.getResolutionPath(from, to);
-		this.rebaser.rebaseAnchors(anchors, over);
-	}
+	public rebaseBranch(
+		source: GraphCommit<TChange>,
+		target: GraphCommit<TChange>,
+		base = target,
+	): [newSource: GraphCommit<TChange>, sourceChange: TChange] {
+		// Get both source and target as path arrays
+		const sourcePath: GraphCommit<TChange>[] = [];
+		const targetPath: GraphCommit<TChange>[] = [];
+		const ancestor = findCommonAncestor([source, sourcePath], [target, targetPath]);
+		assert(ancestor !== undefined, "branch A and branch B must be related");
+
+		// Find where `base` is in the target branch
+		const baseIndex = targetPath.findIndex((r) => r === base);
+		if (baseIndex === -1) {
```

Comment:
I'm having trouble following what valid scenario could land us in this if branch.
Maybe a comment would be good.

Reply:
It happens in the common case, actually, when base === target. Because the paths returned by `findCommonAncestor` do not include the ancestor itself, and when base === target then base === target === ancestor. There is already a comment mentioning this just before the return, but let me know if you think of a better way to state it.

Reply:
Thanks. Wouldn't we be able to change the assert below to `base === target` then?

Reply:
For all correct inputs, yes, but this is also checking the case where somebody passes in a target and base that are just completely unrelated/disjoint, which is invalid input.

Reply:
Actually, yes I think you are right :) That's the only case that is valid. Good catch!

Reply:
Okay, I lied. We forgot about the case where `base` on the target branch, but is _behind_ the place where `source` branches off. In other words, source is already caught up relative to base and there's nothing to do. The assert is checking for that case. We could also change the contract to be "if base and target are unrelated then behavior is undefined" rather than asserting if we're worried about the assert being expensive.

Reply:
Ah, I see. I mostly wanted to test my understanding. The perf improvement was just a nice aside. I'd vote to keep the assert for now.